### PR TITLE
[Durable Objects] Initial DO JS RPC docs

### PR DESCRIPTION
**Pull request by vy-ton**
TODO: 

- [x] Use RPC in counter example
- [x] Rename `Configuration` section, add `Send requests to a Durable Object`
- [x] Update pricing to add RPC methods: https://github.com/cloudflare/cloudflare-docs/pull/13235/

Must merge after https://github.com/cloudflare/cloudflare-docs/pull/13252 which has Workers RPC links

DEFER to next PR:

- [ ] Change all occurrences of `state` to `ctx`
- [ ] Update `Get started` tutorial to RPC (maybe let JS RPC bake for a bit?)

### CODE REVIEW

File: content/durable-objects/best-practices/access-durable-object-from-a-worker.md [hunk 1]
```diff
@@ -1,10 +1,10 @@
 ---
 title: Access a Durable Object from a Worker
 pcx_content_type: concept
-weight: 16
+weight: 1
 ---
 
-# Access a Durable Object from a Worker
+# Access Durable Objects from a Worker
```
Comment (GregBrimble) [inline content/durable-objects/best-practices/access-durable-object-from-a-worker.md]: File name is still `access-durable-object-from-a-worker`. You might want to rename to `access-durable-objects-from-a-worker`.

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 2]
```diff
@@ -1,10 +1,10 @@
 ---
-title: Create Durable Object stubs
+title: Create Durable Object stubs & Send Requests
 pcx_content_type: concept
-weight: 16
+weight: 2
 ---
 
-# Create Durable Object stubs
```
Comment (GregBrimble) [inline content/durable-objects/best-practices/create-durable-object-stubs.md:7]: Same here. Consider renaming file when you change the title.

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 3]
```diff
@@ -43,7 +43,45 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
 
 {{</definitions>}}
 
-## 2. Send HTTP requests
+## Requests
+
+There are several ways to send requests to a Durable Object.
+
+- [Sending HTTP requests](/durable-objects/best-practices/create-durable-object-stubs/#send-http-requests) to a Durable Object's `fetch()` handler was initially the only method of communicating with a Durable Object.
+- Invoking Remote Procedure Call (RPC) methods defined on a Durable Object class as of [compatibility date 2024-04-03](TODO). If handling HTTP requests and responses is not required for your Durable Object, [RPC methods](/workers/runtime-apis/rpc/) provide the developer experience of function calls on an instance of a JavaScript class.
```
Comment (GregBrimble) [inline content/durable-objects/best-practices/create-durable-object-stubs.md]: This reads a bit like a changelog to me. I'd try and rephrase it to be a bit more timeless, if possible. Maybe something like:

```suggestion
- [Sending HTTP requests](/durable-objects/best-practices/create-durable-object-stubs/#send-http-requests) to a Durable Object's `fetch()` handler.
- Invoking [Remote Procedure Call (RPC)](/workers/runtime-apis/rpc/) methods defined on a Durable Object class (available for Workers with a [compatibility date greater than or equal to `2024-04-03`](/workers/configuration/compatibility-dates/#remote-procedure-call-rpc)).
```

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 4]
```diff
@@ -43,7 +43,45 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
...
+- Using the [WebSocket API](durable-objects/reference/websockets/).
+
+### RPC methods
+
+To opt into RPC, a Durable Objects class should extend the built-in type `DurableObject`. Then, public methods on a Durable Objects class are exposed as [RPC methods](/workers/runtime-apis/rpc/), which are callable from a Durable Object stub in a Worker. All RPC calls are [asynchronous](/workers/runtime-apis/rpc/lifecycle), accept & return [serializable types](/workers/runtime-apis/rpc/supported-types), and [propagate exceptions](/workers/runtime-apis/rpc/error-handling) to the caller without a stack trace. See [Workers RPC](/workers/runtime-apis/rpc/supported-types) for complete details.
+
+```ts
+export class Counter extends DurableObject {
```
Comment (GregBrimble) [inline content/durable-objects/best-practices/create-durable-object-stubs.md]: You need the import statement here:
```suggestion
import { DurableObject } from "cloudflare:workers";

export class Counter extends DurableObject {
```

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 5]
```diff
@@ -33,7 +33,7 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
 {{<definitions>}}
 
 - `id` {{<type>}}DurableObjectId{{</type>}}
-  - An ID constructed using `newUniqueId()`, `idFromName()`, or `idFromString()` on this Durable Object namespace.
+  - An ID constructed using `newUniqueId()`, `idFromName()`, or `idFromString()` on this Durable Object namespace, for details see [Access a Durable Object](/durable-objects/best-practices/access-durable-object-from-a-worker/#generate-ids-randomly) .
```
Comment (GregBrimble) [inline content/durable-objects/best-practices/create-durable-object-stubs.md]:
```suggestion
  - An ID constructed using `newUniqueId()`, `idFromName()`, or `idFromString()` on this Durable Object namespace. For details, refer to [Access a Durable Object](/durable-objects/best-practices/access-durable-object-from-a-worker/#generate-ids-randomly) .
```

File: content/durable-objects/examples/build-a-counter.md [hunk 6]
```diff
@@ -39,53 +38,56 @@ export default {
...
+    }
 
+		let count = await resp;
     return new Response(`Durable Object '${name}' count: ${count}`);
   }
 };
 
 // Durable Object
-
 export class Counter {
```
Comment (GregBrimble) [inline content/durable-objects/examples/build-a-counter.md]: You need to extend `DurableObject` here (and import at the top of the file).

File: content/durable-objects/examples/build-a-counter.md [hunk 7]
```diff
@@ -39,53 +38,56 @@ export default {
...
-  constructor(state, env) {
-    this.state = state;
+  constructor(ctx, env) {
+    this.ctx = ctx;
   }
```
Comment (GregBrimble) [inline content/durable-objects/examples/build-a-counter.md]: No need to set `ctx` or `env` in a constructor of a Durable Object once you're extending the `DurableObject` class.

```suggestion
```

File: content/durable-objects/examples/build-a-counter.md [hunk 8]
```diff
@@ -39,53 +38,56 @@ export default {
...
 
-    // You do not have to worry about a concurrent request having modified the value in storage. 
+	async increment(amount = 1) {
+    let value = (await this.ctx.storage.get("value")) || 0;
+    value += amount;
+		// You do not have to worry about a concurrent request having modified the value in storage. 
     // "input gates" will automatically protect against unwanted concurrency. 
     // Read-modify-write is safe. 
-    await this.state.storage.put("value", value);
+    this.ctx.storage.put("value", value);
```
Comment (GregBrimble) [inline content/durable-objects/examples/build-a-counter.md]: I can't remember if we're clever here and if this isn't _strictly_ necessary, or if, indeed, it is needed.. so for the avoidance of my doubt, and the readers':

```suggestion
    await this.ctx.storage.put("value", value);
```

File: content/durable-objects/examples/build-a-counter.md [hunk 9]
```diff
@@ -39,53 +38,56 @@ export default {
...
+    return value;
+  }
 
-    return new Response(value);
+	async decrement(amount = 1) {
+    let value = (await this.ctx.storage.get("value")) || 0;
+    value -= amount;
+    this.ctx.storage.put("value", value);
```
Comment (GregBrimble) [inline content/durable-objects/examples/build-a-counter.md]: Same here:

```suggestion
    await this.ctx.storage.put("value", value);
```

File: content/durable-objects/examples/build-a-counter.md [hunk 10]
```diff
@@ -97,12 +99,13 @@ export class Counter {
 ---
 filename: index.ts
 ---
+import {DurableObject} from "cloudflare:workers";
```
Comment (GregBrimble) [inline content/durable-objects/examples/build-a-counter.md]: Stylistic nit:

```suggestion
import { DurableObject } from "cloudflare:workers";
```

File: content/durable-objects/examples/build-a-counter.md [hunk 11]
```diff
@@ -39,53 +38,56 @@ export default {
 
     // Construct the stub for the Durable Object using the ID. 
     // A stub is a client Object used to send messages to the Durable Object.
-    let obj = env.COUNTER.get(id);
+    let stub = env.COUNTER.get(id);
```
Comment (GregBrimble) [inline content/durable-objects/examples/build-a-counter.md]: Here, we're using `COUNTER` in the JS example, but `COUNTERS` in the TS example. We should be consistent.

File: content/durable-objects/examples/build-a-counter.md [hunk 12]
```diff
@@ -39,53 +38,56 @@ export default {
...
+				break;
+      case "/":
+        // Serves the current value.
+				resp = await stub.getCounterValue();
+        break;
+      default:
+        return new Response("Not found", { status: 404 });
+    }
 
+		let count = await resp;
```
Comment (GregBrimble) [inline content/durable-objects/examples/build-a-counter.md]: You've already awaited the value here, and I think `count` as the return variable name makes more sense than `resp`:

```suggestion
    let count = null;
		switch (url.pathname) {
      case "/increment":
				count = await stub.increment();
        break;
			case "/decrement":
				count = await stub.decrement();
				break;
      case "/":
        // Serves the current value.
				count = await stub.getCounterValue();
        break;
      default:
        return new Response("Not found", { status: 404 });
    }

```

File: content/durable-objects/examples/build-a-counter.md [hunk 13]
```diff
@@ -118,59 +121,59 @@ export default {
...
-    let url = new URL(request.url);
-
-    // Durable Object storage is automatically cached in-memory, so reading the
-    // same key every request is fast. 
-    // You could also store the value in a class member if you prefer.
-    let value: number = (await this.state.storage.get("value")) || 0;
+    let stub = env.COUNTERS.get(id);
 
-    switch (url.pathname) {
+		let resp = null;
```
Comment (GregBrimble) [inline content/durable-objects/examples/build-a-counter.md]: Comments for JS version apply here too.

File: content/workers/_partials/_durable_objects_pricing.md
Comment (GregBrimble) [inline content/workers/_partials/_durable_objects_pricing.md:1]: Just FYI, @elithrar did a slightly different version here: https://github.com/cloudflare/cloudflare-docs/pull/13235/files

File: content/workers/_partials/_durable_objects_pricing.md [hunk 14]
```diff
@@ -12,14 +12,15 @@ _build:
...
+<sup>1</sup> Requests include all incoming HTTP requests, RPC method calls, WebSocket messages, and alarm invocations. A request is needed to create a WebSocket connection. There is no charge for outgoing WebSocket messages, nor for incoming [WebSocket protocol pings](https://www.rfc-editor.org/rfc/rfc6455#section-5.5.2). Billing-only applies a 20:1 ratio to incoming WebSocket messages to factor in smaller messages for real-time communication. For example, 100 WebSocket incoming messages would be charged as 5 requests for billing purposes. The 20:1 ratio does not affect Durable Object metrics and analytics, which reflect actual usage.
```
Comment (GregBrimble) [inline content/workers/_partials/_durable_objects_pricing.md]: Technically, I think it's an RPC _session_ that we'll bill for, rather than method calls. If I'm right, consider linking to https://github.com/cloudflare/cloudflare-docs/pull/13252/files#diff-2f3d01cfb5c442a1cab3aec41e366ce083d59f78defe995a8c586ad96a53ed23R111.

Comment (GregBrimble) [reply content/workers/_partials/_durable_objects_pricing.md]: Also, that "Billing-only applies..." sentence could be clarified a bit, imo. Even just changing "Billing-only applies..." to "For the consideration of billing, we apply..." makes it much more readable.

File: content/_redirects [hunk 15]
```diff
@@ -313,6 +313,8 @@
 /durable-objects/how-to/ /durable-objects/configuration/ 301
 /durable-objects/how-to/access-durable-object-from-a-worker/ /durable-objects/configuration/access-durable-object-from-a-worker/ 301
 /durable-objects/how-to/create-durable-object-stubs/ /durable-objects/configuration/create-durable-object-stubs/ 301
+/durable-objects/configuration/access-durable-object-from-a-worker/ /durable-objects/best-practices/access-durable-object-from-a-worker/ 301
```
Comment (Maddy-Cloudflare) [inline content/_redirects]: What do you mean by "Best practices"? How do you define a "Best practice"? I'd love to know more about this decision.

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 16]
```diff
@@ -1,10 +1,10 @@
 ---
-title: Create Durable Object stubs
+title: Create Durable Object stubs & Send Requests
```
Comment (Maddy-Cloudflare) [inline content/durable-objects/best-practices/create-durable-object-stubs.md]:
```suggestion
title: Create Durable Object stubs and send requests
```

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 17]
```diff
@@ -1,10 +1,10 @@
...
 pcx_content_type: concept
-weight: 16
+weight: 2
 ---
 
-# Create Durable Object stubs
+# Create Durable Object stubs & send requests
```
Comment (Maddy-Cloudflare) [inline content/durable-objects/best-practices/create-durable-object-stubs.md]:
```suggestion
# Create Durable Object stubs and send requests
```

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 18]
```diff
@@ -22,7 +22,7 @@ E-order is a concept deriving from the [E distributed programming language](<htt
 
 {{</Aside>}}
 
-## 1. Obtain a Durable Object stub
+## Obtain a Durable Object stub
```
Comment (Maddy-Cloudflare) [inline content/durable-objects/best-practices/create-durable-object-stubs.md]:
```suggestion
## Get a Durable Object stub
```

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 19]
```diff
@@ -43,7 +43,45 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
 
 {{</definitions>}}
 
-## 2. Send HTTP requests
+## Requests
```
Comment (Maddy-Cloudflare) [inline content/durable-objects/best-practices/create-durable-object-stubs.md]:
```suggestion
## Send requests
```

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 20]
```diff
@@ -43,7 +43,45 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
...
+
+There are several ways to send requests to a Durable Object.
```
Comment (Maddy-Cloudflare) [inline content/durable-objects/best-practices/create-durable-object-stubs.md]:
```suggestion
You can send requests to a Durable Object by:
```

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 21]
```diff
@@ -43,7 +43,45 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
...
+- Invoking Remote Procedure Call (RPC) methods defined on a Durable Object class as of [compatibility date 2024-04-03](TODO). If handling HTTP requests and responses is not required for your Durable Object, [RPC methods](/workers/runtime-apis/rpc/) provide the developer experience of function calls on an instance of a JavaScript class.
+- Using the [WebSocket API](durable-objects/reference/websockets/).
+
+### RPC methods
+
+To opt into RPC, a Durable Objects class should extend the built-in type `DurableObject`. Then, public methods on a Durable Objects class are exposed as [RPC methods](/workers/runtime-apis/rpc/), which are callable from a Durable Object stub in a Worker. All RPC calls are [asynchronous](/workers/runtime-apis/rpc/lifecycle), accept & return [serializable types](/workers/runtime-apis/rpc/supported-types), and [propagate exceptions](/workers/runtime-apis/rpc/error-handling) to the caller without a stack trace. See [Workers RPC](/workers/runtime-apis/rpc/supported-types) for complete details.
```
Comment (Maddy-Cloudflare) [inline content/durable-objects/best-practices/create-durable-object-stubs.md]:
```suggestion
To use RPC, a Durable Objects class should extend the built-in type `DurableObject`. Then, public methods on a Durable Objects class are exposed as [RPC methods](/workers/runtime-apis/rpc/), which you can call from a Durable Object stub in a Worker. All RPC calls are [asynchronous](/workers/runtime-apis/rpc/lifecycle), accept and return [serializable types](/workers/runtime-apis/rpc/supported-types), and [propagate exceptions](/workers/runtime-apis/rpc/error-handling) to the caller without a stack trace. Refer to [Workers RPC](/workers/runtime-apis/rpc/supported-types) for complete details.
```

Comment (Maddy-Cloudflare) [reply content/durable-objects/best-practices/create-durable-object-stubs.md]: Would be nice to add one line on what an RPC method (or methods) is at the beginning of the paragraph.

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 22]
```diff
@@ -43,7 +43,45 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
...
+
+```ts
+// A Worker calling a Durable Object
+let durableObjectStub = env.COUNTERS.get(id);
+let response = await durableObjectStub.increment();
+```
+
+{{<Aside type="note">}}
+
+Historically, the `ctx` object for Durable Objects has been given a variety of different names in example code, most commonly `state`. Until introduction of [RPC methods](/workers/runtime-apis/rpc/), `state` was just the name of a Durable Object constructor parameter, so an application could use whatever name they wanted. With RPC, the `DurableObject` superclass defines these values as properties, so they must have official names. We have standardized on `ctx` for consistency between `DurableObject` and `WorkerEntrypoint`.
```
Comment (Maddy-Cloudflare) [inline content/durable-objects/best-practices/create-durable-object-stubs.md]:
```suggestion
Prior to RPC, the `ctx` object for Durable Objects was known as `state`. `state` was the name of the Durable Object constructor parameter, so an application could use any other name. With RPC, the `DurableObject` superclass defines these values as properties, so they must have official names. The Durable Objects team has chosen to use `ctx` for consistency between `DurableObject` and `WorkerEntrypoint`.
```

File: content/durable-objects/best-practices/create-durable-object-stubs.md [hunk 23]
```diff
@@ -43,7 +43,45 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
...
+
+{{</Aside>}}
+
+See [Build a Counter](/durable-objects/examples/build-a-counter/) for a full example.
```
Comment (Maddy-Cloudflare) [inline content/durable-objects/best-practices/create-durable-object-stubs.md]:
```suggestion
Refer to [Build a Counter](/durable-objects/examples/build-a-counter/) for a full example.
```

File: content/durable-objects/best-practices/create-durable-object-stubs.md (hunk 3 above)
Comment (Maddy-Cloudflare) [reply content/durable-objects/best-practices/create-durable-object-stubs.md]: Since **_Sending HTTP requests_** is at the top of the list, we should start by explaining that, instead of RPC methods.

**Review by Maddy-Cloudflare — CHANGES_REQUESTED** [review]

File: content/durable-objects/best-practices/create-durable-object-stubs.md (hunk 3 above)
Comment (vy-ton) [reply content/durable-objects/best-practices/create-durable-object-stubs.md]: HTTP first is for historical reasons. Eventually, we will default to RPC methods everywhere and should start here to draw developer attention

File: content/workers/_partials/_durable_objects_pricing.md
Comment (vy-ton) [reply content/workers/_partials/_durable_objects_pricing.md:1]: yes, we're going to pull into this PR

File: content/_redirects (hunk 15 above)
Comment (vy-ton) [reply content/_redirects]: `Best Practices` convey to a developer to start here with DO content to build successful apps. Other suggestions have been proposed like `Starter Guide, 101, Basics, Developer Guide`.

"Best Practices" is what I've settled on because we cannot stick with `Configuration` which does not connote build/develop for our audience

File: content/durable-objects/examples/build-a-counter.md [hunk 24]
```diff
@@ -118,59 +119,55 @@ export default {
...
+  }
+
+	async increment(amount = 1) {
+    let value: number = (await this.ctx.storage.get("value")) || 0;
+    value += amount;
+		// You do not have to worry about a concurrent request having modified the value in storage. 
     // "input gates" will automatically protect against unwanted concurrency. 
     // Read-modify-write is safe. 
-    this.state.storage.put("value", value);
+    this.ctx.storage.put("value", value);
```
Comment (GregBrimble) [inline content/durable-objects/examples/build-a-counter.md]:
```suggestion
    await this.ctx.storage.put("value", value);
```

File: content/durable-objects/examples/build-a-counter.md [hunk 25]
```diff
@@ -118,59 +119,55 @@ export default {
...
+    return value;
+  }
 
-    return new Response(value.toString());
+	async decrement(amount = 1) {
+    let value: number = (await this.ctx.storage.get("value")) || 0;
+    value -= amount;
+    this.ctx.storage.put("value", value);
```
Comment (GregBrimble) [inline content/durable-objects/examples/build-a-counter.md]:
```suggestion
    await this.ctx.storage.put("value", value);
```

File: content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md [hunk 26]
```diff
@@ -43,7 +43,47 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
 
 {{</definitions>}}
 
-## 2. Send HTTP requests
+## Send requests
+
+You can send requests to a Durable Object by:
+
+- [Sending HTTP requests](/durable-objects/best-practices/create-durable-object-stubs/#send-http-requests) to a Durable Object's `fetch()` handler.
+- Invoking [Remote Procedure Call (RPC)](/workers/runtime-apis/rpc/) methods defined on a Durable Object class (available for Workers with a [compatibility date greater than or equal to `2024-04-03`](/workers/configuration/compatibility-dates/#remote-procedure-call-rpc)).
```
Comment (irvinebroque) [inline content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md:51]: Tricky to frame but technically when you call an RPC on a DO, you are not sending a request

File: content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md [hunk 27]
```diff
@@ -43,7 +43,47 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
...
+
+```ts
+// A Worker calling a Durable Object
+let durableObjectStub = env.COUNTERS.get(id);
+let response = await durableObjectStub.increment();
+```
+
+{{<Aside type="note">}}
+
+Prior to [RPC introduction](/workers/configuration/compatibility-dates/#remote-procedure-call-rpc), the `ctx` object for Durable Objects was known as `state`. `state` was the name of the Durable Object constructor parameter, so an application could use any other name. With RPC, the `DurableObject` superclass defines these values as properties, so they must have official names. `ctx` was chosen for consistency between `DurableObject` and `WorkerEntrypoint`.
```
Comment (irvinebroque) [inline content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md]:
```suggestion
When you write a Worker that does not extend the [`DurableObject` class](/workers/configuration/compatibility-dates/#remote-procedure-call-rpc), the `ctx` object for Durable Objects was known as `state`. `state` was the name of the Durable Object constructor parameter, so an application could use any other name. With RPC, the `DurableObject` superclass defines these values as properties, so they must have official names. `ctx` was chosen for consistency between `DurableObject` and `WorkerEntrypoint`.
```

it's this, right? that it still works the same way as before if you don't extend from DurableObject

File: content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md (hunk 26 above)
Comment (vy-ton) [reply content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md:51]: more context [here](https://chat.google.com/room/AAAAVrF5vCc/rHX2avbTVOI/YUTgOkqMAmw?cls=10), will work to clarify

File: content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md (hunk 27 above)
Comment (vy-ton) [reply content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md]: added suggestion and tweaked wording some more

File: content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md [hunk 28]
```diff
@@ -43,7 +43,47 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
...
+
+```ts
+// A Worker calling a Durable Object
+let durableObjectStub = env.COUNTERS.get(id);
+let response = await durableObjectStub.increment();
+```
+
+{{<Aside type="note">}}
+
+When you write a Worker that does not extend the [`DurableObject` class](/workers/configuration/compatibility-dates/#remote-procedure-call-rpc), the `ctx` object for Durable Objects was known as `state` in code examples. `state` was the name of the Durable Object constructor parameter, so an application could use any other name. With RPC, the `DurableObject` superclass defines these values as properties, so they must have official names. `ctx` was chosen for consistency between `DurableObject` and `WorkerEntrypoint`.
```
Comment (irvinebroque) [inline content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md]:
```suggestion
When you write a Worker that does not extend the [`DurableObject` class](/workers/configuration/compatibility-dates/#remote-procedure-call-rpc), the `ctx` object in your Durable Object is instead called `state`, and is provided as the first argument to the constructor, but not set as a property of your class.

With RPC, the `DurableObject` superclass defines `ctx` and `env` as class properties. What was previously called `state` is now called `ctx` when you extend the `DurableObject` class.

This naming change from `state` to `ctx` was made both to ensure consistency between `DurableObject` and `WorkerEntrypoint`, and to avoid setting the false expectation that directly mutating the value of `state` would persist data.
```

@vy-ton — feel free to take or leave any of this. Tricky one to word, just taking a shot

File: content/durable-objects/examples/build-a-counter.md [hunk 29]
```diff
@@ -9,7 +9,7 @@ weight: 3
 layout: example
 ---
 
-This example shows how to build a counter using Durable Objects and Workers that can print, increment, and decrement a `name` provided by the URL query string parameter, for example, `?name=A`.
+This example shows how to build a counter using Durable Objects and Workers with RPC methods that can print, increment, and decrement a `name` provided by the URL query string parameter, for example, `?name=A`.
```
Comment (irvinebroque) [inline content/durable-objects/examples/build-a-counter.md]:
```suggestion
This example shows how to build a counter using Durable Objects and Workers with [RPC methods](/workers/runtime-apis/rpc) that can print, increment, and decrement a `name` provided by the URL query string parameter, for example, `?name=A`.
```

File: content/workers/_partials/_durable_objects_pricing.md [hunk 30]
```diff
@@ -12,14 +12,17 @@ _build:
 |          | Paid plan                                         |
 | -------- | ------------------------------------------------- |
 | HTTP requests<sup>1</sup>               | 1 million, + $0.15/million                        |
+| RPC methods                             | 1 million, + $0.15/million                        |
```
Comment (irvinebroque) [inline content/workers/_partials/_durable_objects_pricing.md]:
```suggestion
| RPC sessions                             | 1 million, + $0.15/million                        |
```

+ link to content/workers/runtime-apis/rpc/lifecycle.md from other PR?

(think we need crisp definition of what a session is)

File: content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md (hunk 27 above)
Comment (geelen) [reply content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md]: Yeah. Also if you really want, you can do:

```js
export class Counter extends DurableObject {
  state = this.ctx

  // rest of your code here
}
```

It's just that you'll end up with both `this.state` and `this.ctx` pointing to the same thing

**Review by Maddy-Cloudflare — APPROVED** [review]

File: content/_redirects (hunk 15 above)
Comment (Maddy-Cloudflare) [reply content/_redirects]: I strongly disagree (respectfully). But you can go ahead. :)

File: content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md (hunk 27 above)
Comment (vy-ton) [reply content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md]: Separately, we will replace all mentions of `state` with `ctx`. @Maddy-Cloudflare are you able to tackle this next week?

File: content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md (hunk 26 above)
Comment (irvinebroque) [reply content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md:51]: looks great

File: content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md [hunk 31]
```diff
@@ -43,7 +43,47 @@ let durableObjectStub = OBJECT_NAMESPACE.get(id);
...
+
+You can call a Durable Object by:
+
+- [Sending HTTP requests](/durable-objects/best-practices/create-durable-object-stubs-and-send-requests/#send-http-requests) to a Durable Object's `fetch()` handler.
+- Invoking [Remote Procedure Call (RPC)](/workers/runtime-apis/rpc/) methods defined on a Durable Object class (available for Workers with a [compatibility date greater than or equal to `2024-04-03`](/workers/configuration/compatibility-dates/#remote-procedure-call-rpc)).
+- Using the [WebSocket API](/durable-objects/reference/websockets/).
+
+### Call RPC methods
+
+To use RPC, a Durable Objects class should extend the built-in type `DurableObject`. Then, public methods on a Durable Objects class are exposed as [RPC methods](/workers/runtime-apis/rpc/), which you can call from a Durable Object stub in a Worker. All RPC calls are [asynchronous](/workers/runtime-apis/rpc/lifecycle), accept and return [serializable types](/workers/runtime-apis/rpc/supported-types), and [propagate exceptions](/workers/runtime-apis/rpc/error-handling) to the caller without a stack trace. Refer to [Workers RPC](/workers/runtime-apis/rpc/supported-types) for complete details.
```
Comment (irvinebroque) [inline content/durable-objects/best-practices/create-durable-object-stubs-and-send-requests.md]:
```suggestion
To use RPC, a Durable Objects class must extend the built-in type `DurableObject`. Then, public methods on a Durable Objects class are exposed as [RPC methods](/workers/runtime-apis/rpc/), which you can call from a Durable Object stub in a Worker. All RPC calls are [asynchronous](/workers/runtime-apis/rpc/lifecycle), accept and return [serializable types](/workers/runtime-apis/rpc/supported-types), and [propagate exceptions](/workers/runtime-apis/rpc/error-handling) to the caller without a stack trace. Refer to [Workers RPC](/workers/runtime-apis/rpc/supported-types) for complete details.
```

**Review by irvinebroque — APPROVED** [review]

File: content/workers/_partials/_durable_objects_pricing.md (hunk 30 above)
Comment (vy-ton) [reply content/workers/_partials/_durable_objects_pricing.md]: @irvinebroque Are developers familiar with RPC sessions? Worried developers will struggle to understand RPC sessions in context of DO request billing. For the common case, calling a RPC method on a DO stub is 1 billed request.

I know there's a more advanced use case with `RpcTarget` where subsequent calls on a RPC session are not counted as billed invocations.

Comment (vy-ton) [reply content/workers/_partials/_durable_objects_pricing.md]: took another stab at explain this on the DO pricing page

File: content/workers/_partials/_durable_objects_pricing.md [hunk 32]
```diff
@@ -11,15 +11,19 @@ _build:
...
+| Duration<sup>3</sup>                    | 400,000 GB-s, + $12.50/million GB-s<sup>4,5</sup> |
 
 {{</table-wrap>}}
 
-<sup>1</sup> A request is needed to create a WebSocket connection. There is no charge for outgoing WebSocket messages, nor for incoming [WebSocket protocol pings](https://www.rfc-editor.org/rfc/rfc6455#section-5.5.2). For compute requests billing-only, a 20:1 ratio is applied to incoming WebSocket messages to factor in smaller messages for real-time communication. For example, 100 WebSocket incoming messages would be charged as 5 requests for billing purposes. The 20:1 ratio does not affect Durable Object metrics and analytics, which reflect actual usage.
+<sup>1</sup> A [RPC session](/workers/runtime-apis/rpc/lifecycle/) represents a billed request. Every [RPC method call](/durable-objects/best-practices/create-durable-object-stubs-and-send-requests/#call-rpc-methods) on a [Durable Objects stub](/durable-objects/best-practices/create-durable-object-stubs-and-send-requests/#get-a-durable-object-stub) is its own RPC session and therefore a single billed request.
```
Comment (irvinebroque) [inline content/workers/_partials/_durable_objects_pricing.md]:
```suggestion
<sup>1</sup> Each [RPC session](/workers/runtime-apis/rpc/lifecycle/) is billed as one request to your Durable Object. Every [RPC method call](/durable-objects/best-practices/create-durable-object-stubs-and-send-requests/#call-rpc-methods) on a [Durable Objects stub](/durable-objects/best-practices/create-durable-object-stubs-and-send-requests/#get-a-durable-object-stub) is its own RPC session and therefore a single billed request.
```

Think that this needs to be tweaked a bit based on content below?

Since only that top-level calls to create new stubs are billed as requests?

i.e.:

```js
using foo = await env.MY_DO.bar(); // billed as a request
await foo.baz(); // treated as part of the same session, not billed
```

^ where both of these calls would be considered calls on a stub from a Durable Object?

I wonder if one way to approach would be something like:

"Every time you call a public class method of your Durable Object via RPC, this creates a session, and each session is billed as one request"

@kentonv for correctness ^ — accurate?

File: content/workers/_partials/_durable_objects_pricing.md [hunk 33]
```diff
@@ -11,15 +11,19 @@ _build:
...
 
-<sup>2</sup> Application level auto-response messages handled by [`state.setWebSocketAutoResponse()`](/durable-objects/api/websockets/) will not incur additional wall-clock time, and so they will not be charged.
+RPC method calls can return objects (stubs) extending [`RpcTarget`](/workers/runtime-apis/rpc/lifecycle/#lifetimes-memory-and-resource-management) and invoke calls on those stubs. Subsequent calls on the returned stub are part of the same RPC session and are not billed as separate requests.
```
Comment (irvinebroque) [inline content/workers/_partials/_durable_objects_pricing.md]: Maybe this is where to put code example like:

```suggestion
RPC method calls can return objects (stubs) extending [`RpcTarget`](/workers/runtime-apis/rpc/lifecycle/#lifetimes-memory-and-resource-management) and invoke calls on those stubs. Subsequent calls on the returned stub are part of the same RPC session and are not billed as separate requests. For example:

```js
using foo = await env.MY_DO.bar(); // billed as a request
await foo.baz(); // treated as part of the same session, not billed as a request
```

```

File: content/workers/_partials/_durable_objects_pricing.md (hunk 32 above)
Comment (vy-ton) [reply content/workers/_partials/_durable_objects_pricing.md]: > ^ where both of these calls would be considered calls on a stub from a Durable Object?

but the second line `await foo.baz()` is not a call on the Durable Object stub itself. That's the difference to highlight.

File: content/workers/_partials/_durable_objects_pricing.md (hunk 33 above)
Comment (vy-ton) [reply content/workers/_partials/_durable_objects_pricing.md]: +1 code example, proposed edits:

```js
let durableObjectStub = OBJECT_NAMESPACE.get(id); // retrieve Durable Object stub
using foo = await durableObjectStub.bar(); // billed as a request
await foo.baz(); // treated as part of the same RPC session created by calling bar(), not billed as a request
await durableObjectStub.cat(); // billed as a request

Comment (irvinebroque) [reply content/workers/_partials/_durable_objects_pricing.md]: ship it!

File: content/workers/_partials/_durable_objects_pricing.md (hunk 32 above)
Comment (irvinebroque) [reply content/workers/_partials/_durable_objects_pricing.md]: yeah, i think we are agreeing, just trying to figure out how to clarify what "is a Durable Object stub" i.e. I read that as somewhat ambiguous, could mean either:

1. Any stub returned by a Durable Object
2. Specifically, the stub "of" the Durable Object itself (in contrast to the stub provided by an `RpcTarget`

Think what you have probably fine since clarified below? Was just wondering about if way to be super explicit
